pkg/controllers/nodeipam/ipam/cidr_allocator.go

```diff
@@ -0,0 +1,480 @@
+		}
+		allocated.allocatedCIDRs[idx] = podCIDR
+	}
+
+	//queue the assignment
+	klog.V(4).Infof("Putting node %s with CIDR %v into the work queue", node.Name, allocated.allocatedCIDRs)
+	r.nodeCIDRUpdateChannel <- allocated
+	return nil
+}
+
+// ReleaseCIDR marks node.podCIDRs[...] as unused in our tracked cidrSets
+func (r *rangeAllocator) ReleaseCIDR(node *v1.Node) error {
+	if node == nil || len(node.Spec.PodCIDRs) == 0 {
+		return nil
+	}
+
+	for idx, cidr := range node.Spec.PodCIDRs {
+		_, podCIDR, err := net.ParseCIDR(cidr)
+		if err != nil {
+			return fmt.Errorf("failed to parse CIDR %s on Node %v: %v", cidr, node.Name, err)
+		}
+
+		// If node has a pre allocate cidr that does not exist in our cidrs.
+		// This will happen if cluster went from dualstack(multi cidrs) to non-dualstack
+		// then we have now way of locking it
+		if idx >= len(r.cidrSets) {
+			return fmt.Errorf("node:%s has an allocated cidr: %v at index:%v that does not exist in cluster cidrs configuration", node.Name, cidr, idx)
+		}
+
+		klog.V(4).Infof("release CIDR %s for node:%v", cidr, node.Name)
+		if err = r.cidrSets[idx].Release(podCIDR); err != nil {
+			return fmt.Errorf("error when releasing CIDR %v: %v", cidr, err)
+		}
+	}
+	return nil
+}
+
+// updateCIDRsAllocation assigns CIDR to Node and sends an update to the API server.
+func (r *rangeAllocator) updateCIDRsAllocation(data NodeReservedCIDRs) error {
+	var err error
+	var node *v1.Node
+	defer r.removeNodeFromProcessing(data.nodeName)
+	cidrsString := cidrsAsString(data.allocatedCIDRs)
+	node, err = r.nodeLister.Get(data.nodeName)
+	if err != nil {
+		klog.Errorf("Failed while getting node %v for updating Node.Spec.PodCIDRs: %v", data.nodeName, err)
+		return err
+	}
+
+	// if cidr list matches the proposed.
+	// then we possibly updated this node
+	// and just failed to ack the success.
+	if len(node.Spec.PodCIDRs) == len(data.allocatedCIDRs) {
+		match := true
+		for idx, cidr := range cidrsString {
+			if node.Spec.PodCIDRs[idx] != cidr {
+				match = false
+				break
+			}
+		}
+		if match {
+			klog.V(4).Infof("Node %v already has allocated CIDR %v. It matches the proposed one.", node.Name, data.allocatedCIDRs)
+			return nil
+		}
+	}
+
+	// node has cidrs, release the reserved
+	if len(node.Spec.PodCIDRs) != 0 {
+		klog.Errorf("Node %v already has a CIDR allocated %v. Releasing the new one.", node.Name, node.Spec.PodCIDRs)
+		for idx, cidr := range data.allocatedCIDRs {
+			if releaseErr := r.cidrSets[idx].Release(cidr); releaseErr != nil {
+				klog.Errorf("Error when releasing CIDR idx:%v value: %v err:%v", idx, cidr, releaseErr)
+			}
+		}
+		return nil
+	}
+
+	//  fetch ipv6 cidr address
+	if node.Spec.ProviderID == "" {
+		klog.Infof("Node %q has empty provider ID", node.Name)
+		return nil
+	}
+
+	// aws:///eu-central-1a/i-07577a7bcf3e576f2
+	instanceID, _ := awsv1.KubernetesInstanceID(node.Spec.ProviderID).MapToAWSInstanceID()
```

Comment:
the error is ignored..what id the instance id is a non-valid instanceid? like a fargate task?